blockchain/skipchain/block.go

```diff
@@ -176,37 +212,133 @@ func (fl ForwardLink) computeHash() ([]byte, error) {
 	return h.Sum(nil), nil
 }
 
+// Chain is an ordered list of skipblocks.
+type Chain []SkipBlock
+
+// GetProof returns a verifiable proof to the latest block of the chain.
+func (c Chain) GetProof() Proof {
+	chain := c[:len(c)-1]
+
+	proof := Proof{
+		ForwardLinks: make([]ForwardLink, len(chain)),
+	}
+
+	for i, block := range chain {
+		proof.ForwardLinks[i] = block.ForwardLinks[0]
+	}
+
+	return proof
+}
+
+// Proof is a chain of forward links to a specific block so that the chain
+// can be followed and verified.
+type Proof struct {
+	GenesisBlock SkipBlock
+	ForwardLinks []ForwardLink
+}
+
+// Pack implements the Packable interface to encode the proof into a protobuf
+// message.
+func (p Proof) Pack() (proto.Message, error) {
+	proof := &ProofProto{
+		ForwardLinks: make([]*ForwardLinkProto, len(p.ForwardLinks)),
+	}
+
+	for i, link := range p.ForwardLinks {
+		packed, err := link.Pack()
+		if err != nil {
+			return nil, xerrors.Errorf("couldn't pack the forward link: %v", err)
+		}
+
+		proof.ForwardLinks[i] = packed.(*ForwardLinkProto)
+	}
+
+	return proof, nil
+}
+
+// Verify follows the chain from the beginning to insure the integrity.
+func (p Proof) Verify(v crypto.Verifier, block SkipBlock) error {
+	pubkeys := p.GenesisBlock.Roster.GetPublicKeys()
+
+	if !bytes.Equal(p.ForwardLinks[0].From, p.GenesisBlock.hash) {
+		fabric.Logger.Error().Msgf("%x != %x", p.ForwardLinks[0].From, p.GenesisBlock.hash)
+		return xerrors.New("mismatching genesis hash in forward link")
+	}
+
+	for i, link := range p.ForwardLinks {
+		if i > 0 {
+			prev := p.ForwardLinks[i-1].To
+			if !bytes.Equal(prev, link.From) {
+				return xerrors.New("mismatching previous target in forward link")
+			}
+		}
+
+		err := link.Verify(v, pubkeys)
+		if err != nil {
+			return xerrors.Errorf("couldn't verify the forward link: %v", err)
+		}
```

Comment:
```suggestion
		err := link.Verify(v, pubkeys)
		if err != nil {
			return xerrors.Errorf("couldn't verify the forward link: %v", err)
		}
		
		if i == 0 {
			continue
		}
		
		prev := p.ForwardLinks[i-1].To
		if !bytes.Equal(prev, link.From) {
			return xerrors.New("mismatching previous target in forward link. "+
				"Expected (prev) %x to equal (from) %x", prev, link.from)
		}
```

Reply:
https://softwareengineering.stackexchange.com/questions/58237/are-break-and-continue-bad-programming-practices

I'm allergic to continue and break for this kind of reasons.

Reply:
But I agree on the ordering of the verification. Same remark about the error message. Let's talk about this tomorrow.